### ISSUE TITLE
Dangling pointer

### ISSUE THREAD

**Issue by importcjj** [vlang/v#740]
**V version: V 0.1.7** 
**OS:x86_64-apple-darwin18.0.0** 
**C compiler version:Apple LLVM version 10.0.0 (clang-1000.10.44.4)** 

**What did you do?**

```go
struct Point {
mut:
    x int
    y int
}


fn max_point(a, b Point) &Point {
    if a.x < b.x {
        return &b
    }
    return &a
}


fn main() {
    a := Point{ x: 1, y: 2}
    mut max := &a
    println(a)
    {
        b := Point{ x: 3, y: 4}
        println(b)
        max = max_point(a, b)
        println(max.x)
    }
    println(max)
    println(max.x)
    println('Hello, World!')
}
```

**What did you expect to see?**
Maybe fail to compile?
 
**What did you see instead?**

```
{ x: 1, y: 2 }
{ x: 3, y: 4 }
3
0x7ffeee317788
-298748024
Hello, World!
```

**Comment by Bowero** [comment]
This problem has increased:

```
[1]    10405 segmentation fault (core dumped)  ./test
```

**Comment by nedpals** [comment]
On the new parser, it now emits an error from the C compiler so maybe the situation has been improved?

```
root@DESKTOP-7SD02HN:~/code# v -b v2 run test.v
compile2()
build_c(/root/.cache/v/test.tmp.c)
g_m2_buf
g_m2_ptr
total_m
nr_mallocs
g_ustring_runes
PARSE: 58ms
CHECK: 7ms
cgen.expr(): bad node v_dot_ast__Expr
cgen.expr(): bad node v_dot_ast__Expr
C GEN: 20ms
cgen done
==================
/root/.cache/v/test.tmp.c: In function ‘main’:
/root/.cache/v/test.tmp.c:4521:26: error: incompatible type for argument 1 of ‘Point_str’
  string tmp1 = Point_str(a); println(tmp1); string_free(tmp1); //MEM2 Point
                          ^
/root/.cache/v/test.tmp.c:811:8: note: expected ‘Point * {aka struct Point *}’ but argument is of type ‘Point {aka struct Point}’
 string Point_str(Point* x) { return tos3("TODO_str"); }
        ^
/root/.cache/v/test.tmp.c:4528:27: error: incompatible type for argument 1 of ‘Point_str’
   string tmp2 = Point_str(b); println(tmp2); string_free(tmp2); //MEM2 Point
                           ^
/root/.cache/v/test.tmp.c:811:8: note: expected ‘Point * {aka struct Point *}’ but argument is of type ‘Point {aka struct Point}’
 string Point_str(Point* x) { return tos3("TODO_str"); }
...
==================
(Use `v -cg` to print the entire error message)

V error: C error.

Please make sure that:
- You have all V dependencies installed.
- You did not declare a C function that was not included. (Try commenting your code that involves C interop)
- You are running the latest version of V. (Try running `v up` and rerunning your command)

If you're confident that all of the above is true, please try running V with the `-cg` option which enables more debugging capabilities.
```

**Comment by emily33901** [comment]
Becuase a and b are brought into the new stack frame the address of them is within that stack frame, when the function returns it is indeed dangling. 

There are 2 ways that I can think of dealing with this
1. by preventing someone from just returning a reference to a parameter or a stack allocated variable (probably the best option)
2. heap allocate the return value so that is guarenteed to be safe (this has the downside of diverging the param and the value it was created from)

**Comment by justinas** [comment]
Still trivial to reproduce.

```vlang
struct User {
    name string
    age &int
}

fn create_user(age int) User {
    user := User{"john", &age}
    return user
}

fn main() {
    user := create_user(13)
    create_user(14)
    println(user.age)
    println(u64(user.age))
}
```

Output:
```
&14
140725022070704
```

**Comment by crthpl** [comment]
Fixed with #9873
